File: telegram-bot/bot.py
#!/usr/bin/env python3
"""
Telegram Bot for Savdo Kalkulyatori Mini App
"""

import os
import logging
from typing import Optional
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup, WebAppInfo
from telegram.ext import Application, CommandHandler, ContextTypes

# Enable logging
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    level=logging.INFO
)
logger = logging.getLogger(__name__)

# Bot configuration
BOT_TOKEN = os.getenv('BOT_TOKEN', 'YOUR_BOT_TOKEN_HERE')
WEBAPP_URL = os.getenv('WEBAPP_URL', 'https://muminoff.com/retail-calc/')

async def start(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Send a message with a button to launch the Mini App."""
    user = update.effective_user
    
    # Create the Mini App button
    keyboard = [
        [
            InlineKeyboardButton(
                text="🧮 Kalkulyatorni ochish",
                web_app=WebAppInfo(url=WEBAPP_URL)
            )
        ]
    ]
    reply_markup = InlineKeyboardMarkup(keyboard)
    
    await update.message.reply_text(
        f"Assalomu alaykum, {user.first_name}! 👋\n\n"
        "Savdo Kalkulyatori Mini App-ga xush kelibsiz!\n\n"
        "Bu bot orqali siz:\n"
        "• Mahsulot narxini hisoblashingiz\n"
        "• Valyuta kursini ko'rishingiz\n"
        "• DataMatrix kod yaratishingiz mumkin\n\n"
        "Boshlash uchun quyidagi tugmani bosing:",
        reply_markup=reply_markup
    )

async def help_command(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Send a message when the command /help is issued."""
    help_text = """
🔹 *Yordam*

*Asosiy buyruqlar:*
/start - Botni ishga tushirish va kalkulyatorni ochish
/help - Yordam olish
/about - Bot haqida ma'lumot

*Kalkulyator xususiyatlari:*
• Asl narxni belgilash (KRW)
• Og'irlik bo'yicha pochta xarajatini hisoblash
• Foyda foizini sozlash
• Real vaqtda valyuta kursini olish
• DataMatrix kod yaratish

*Muammo bo'lsa:* @your_username
    """
    
    await update.message.reply_text(
        help_text, 
        parse_mode='Markdown'
    )

async def about(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Send information about the bot."""
    about_text = """
📊 *Savdo Kalkulyatori*

Versiya: 1.0.0
Yaratilgan: 2025

Bu bot savdogarlar uchun narxlarni hisoblashda yordam beradi.

• Real valyuta kurslari
• Avtomatik narx hisoblash
• DataMatrix kod generatsiyasi

Web versiya: https://muminoff.com/retail-calc/
    """
    
    await update.message.reply_text(
        about_text, 
        parse_mode='Markdown'
    )

async def error_handler(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    """Log errors caused by updates."""
    logger.warning('Update "%s" caused error "%s"', update, context.error)

def main() -> None:
    """Start the bot."""
    # Check if token is set
    if BOT_TOKEN == 'YOUR_BOT_TOKEN_HERE':
        logger.error("Please set your bot token in BOT_TOKEN environment variable!")
        return
    
    # Create the Application
    application = Application.builder().token(BOT_TOKEN).build()

    # Register command handlers
    application.add_handler(CommandHandler("start", start))
    application.add_handler(CommandHandler("help", help_command))
    application.add_handler(CommandHandler("about", about))

    # Register error handler
    application.add_error_handler(error_handler)

    # Run the bot until the user presses Ctrl-C
    logger.info("Starting bot...")
    application.run_polling(allowed_updates=Update.ALL_TYPES)

if __name__ == '__main__':
    main()

File: telegram-bot-node/bot.js
const { Telegraf, Markup } = require('telegraf');
require('dotenv').config();

// Bot configuration
const BOT_TOKEN = process.env.BOT_TOKEN || 'YOUR_BOT_TOKEN_HERE';
const WEBAPP_URL = process.env.WEBAPP_URL || 'https://muminoff.com/retail-calc/';

// Check if token is provided
if (BOT_TOKEN === 'YOUR_BOT_TOKEN_HERE') {
  console.error('Please set your bot token in BOT_TOKEN environment variable!');
  process.exit(1);
}

// Create bot instance
const bot = new Telegraf(BOT_TOKEN);

// Start command
bot.start((ctx) => {
  const userName = ctx.from.first_name || 'Foydalanuvchi';
  
  ctx.reply(
    `Assalomu alaykum, ${userName}! 👋\n\n` +
    `Savdo Kalkulyatori Mini App-ga xush kelibsiz!\n\n` +
    `Bu bot orqali siz:\n` +
    `• Mahsulot narxini hisoblashingiz\n` +
    `• Valyuta kursini ko'rishingiz mumkin\n\n` +
    `Boshlash uchun quyidagi tugmani bosing:`,
    Markup.inlineKeyboard([
      [Markup.button.webApp('🧮 Kalkulyatorni ochish', WEBAPP_URL)]
    ])
  );
});

// Help command
bot.help((ctx) => {
  ctx.replyWithMarkdown(`
🔹 *Yordam*

*Asosiy buyruqlar:*
/start - Botni ishga tushirish va kalkulyatorni ochish
/help - Yordam olish
/about - Bot haqida ma'lumot

*Kalkulyator xususiyatlari:*
• Asl narxni belgilash (KRW)
• Og'irlik bo'yicha pochta xarajatini hisoblash
• Foyda foizini sozlash
• Real vaqtda valyuta kursini olish

*Muammo bo'lsa:* @your_username
  `);
});

// About command
bot.command('about', (ctx) => {
  ctx.replyWithMarkdown(`
📊 *Savdo Kalkulyatori*

Versiya: 1.0.0
Yaratilgan: 2025

Bu bot savdogarlar uchun narxlarni hisoblashda yordam beradi.

• Real valyuta kurslari
• Avtomatik narx hisoblash

Web versiya: ${WEBAPP_URL}
  `);
});

// Handle Mini App data (when Mini App sends data back)
bot.on('web_app_data', (ctx) => {
  const data = ctx.webAppData.data;
  console.log('Received data from Mini App:', data);
  
  // You can process the data here
  ctx.reply(`Ma'lumotlar qabul qilindi: ${data}`);
});

// Error handling
bot.catch((err, ctx) => {
  console.error('Bot error:', err);
  ctx.reply('Xatolik yuz berdi. Iltimos qayta urinib ko\'ring.');
});

// Launch bot
bot.launch().then(() => {
  console.log('Bot started successfully!');
  console.log('Mini App URL:', WEBAPP_URL);
});

// Enable graceful stop
process.once('SIGINT', () => bot.stop('SIGINT'));
process.once('SIGTERM', () => bot.stop('SIGTERM'));